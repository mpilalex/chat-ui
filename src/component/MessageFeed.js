import React from "react";
import Component from "./BaseComponent";
import PropTypes from "prop-types";
import {Feed, FeedContent, FeedEvent, FeedSummary, FeedLabel, FeedDate, FeedExtra} from "semantic-ui-react";
import _ from "underscore";
import moment from "moment"

class MessageFeed extends Component {

    componentDidUpdate(prevProps){
        if(this.props.messages.length !== prevProps.messages.lenght){
            this.refs.messageFeed.scrollTop = this.refs.messageFeed.scrollHeight;
        }
    }

    render() {
        let {messages, clientId} = this.props;
        messages = _.sortBy(messages, "timestamp");
        messages.reverse();
        return <div className={"message-feed"} ref={"messageFeed"}>
            <Feed>
                {
                    _.map(messages, message => {
                        const className = message.clientId === clientId ? "mine" : "user";
                        return <FeedEvent className={className} key={message.id}>
                            <FeedLabel>
                                <img src={message.avatar}/>
                            </FeedLabel>
                            <FeedContent>
                                <FeedSummary>
                                    {message.username}
                                    <FeedDate>{moment(message.timestamp).fromNow()}</FeedDate>
                                </FeedSummary>
                                <FeedExtra text>
                                    {message.text}
                                </FeedExtra>
                            </FeedContent>
                        </FeedEvent>
                    })
                }
            </Feed>
        </div>
    }
}

MessageFeed.propTypes = {
    messages: PropTypes.arrayOf(PropTypes.object),
    clientId: PropTypes.string.isRequired
};


export default MessageFeed;