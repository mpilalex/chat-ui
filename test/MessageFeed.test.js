import React from 'react';
import MessageFeed from '../src/component/MessageFeed';
import renderer from 'react-test-renderer';
import {Provider} from "react-redux";
import reduxStore from "./../src/store";
import avatars from "../src/assets/avatars.json";

const messages = [
    {
        id: "id1",
        clientId:"clientId1",
        text: "Hi",
        avatar: avatars[0],
        username:"user1",
        timestamp: Date.now()
    },
    {
        id: "id2",
        clientId:"clientId2",
        text: "Hi",
        avatar: avatars[0],
        username:"user2",
        timestamp: Date.now() - 3600
    }
];

test('username render', () => {

    const messageFeedNoMessages = renderer.create(<Provider store={reduxStore}><MessageFeed clientId={"clientId1"} messages={[]} /></Provider>);
    const usernameFeedNoMessagesTree = messageFeedNoMessages.toJSON();
    expect(usernameFeedNoMessagesTree).toMatchSnapshot();

    const messageFeedOneMessage = renderer.create(<Provider store={reduxStore}><MessageFeed clientId={"clientId1"} messages={[messages[0]]} /></Provider>);
    const usernameFeedOneMessageTree = messageFeedOneMessage.toJSON();
    expect(usernameFeedOneMessageTree).toMatchSnapshot();

    const messageFeedMultipleMessage = renderer.create(<Provider store={reduxStore}><MessageFeed clientId={"clientId1"} messages={messages} /></Provider>);
    const usernameFeedMultipleMessagesTree = messageFeedMultipleMessage.toJSON();
    expect(usernameFeedMultipleMessagesTree).toMatchSnapshot();

});
