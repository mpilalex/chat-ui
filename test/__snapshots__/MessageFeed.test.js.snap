// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`username render 1`] = `
<div
  className="message-feed"
>
  <div
    className="ui feed"
  />
</div>
`;

exports[`username render 2`] = `
<div
  className="message-feed"
>
  <div
    className="ui feed"
  >
    <div
      className="event mine"
    >
      <div
        className="label"
      >
        <img
          src="https://spotim-demo-chat-server.herokuapp.com/avatars/001-snorlax.png"
        />
      </div>
      <div
        className="content"
      >
        <div
          className="summary"
        >
          user1
          <div
            className="date"
          >
            a few seconds ago
          </div>
        </div>
        <div
          className="text extra"
        >
          Hi
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`username render 3`] = `
<div
  className="message-feed"
>
  <div
    className="ui feed"
  >
    <div
      className="event mine"
    >
      <div
        className="label"
      >
        <img
          src="https://spotim-demo-chat-server.herokuapp.com/avatars/001-snorlax.png"
        />
      </div>
      <div
        className="content"
      >
        <div
          className="summary"
        >
          user1
          <div
            className="date"
          >
            a few seconds ago
          </div>
        </div>
        <div
          className="text extra"
        >
          Hi
        </div>
      </div>
    </div>
    <div
      className="event user"
    >
      <div
        className="label"
      >
        <img
          src="https://spotim-demo-chat-server.herokuapp.com/avatars/001-snorlax.png"
        />
      </div>
      <div
        className="content"
      >
        <div
          className="summary"
        >
          user2
          <div
            className="date"
          >
            a few seconds ago
          </div>
        </div>
        <div
          className="text extra"
        >
          Hi
        </div>
      </div>
    </div>
  </div>
</div>
`;
